### PR TITLE
LIB: Add mindspore backend

### PR DESCRIPTION
**Pull request by chou-shun**
1. What this PR does: Add mindspore backend
2. Special notes for your reviewer: If both GPU and NPU exist in the environment, NPU is preferred

Signed-off-by: zhangjun <zhangjun286@huawei.com>

### CODE REVIEW

**Comment by chou-shun** [comment]
> 1. Add mindspore backend
> 2. If both GPU and NPU exist in the environment, NPU is preferred
> 
> Signed-off-by: zhangjun [zhangjun286@huawei.com](mailto:zhangjun286@huawei.com)

File: lib/sedna/backend/__init__.py [hunk 1]
```diff
@@ -27,14 +27,20 @@ def set_backend(estimator=None, config=None):
...
         'BACKEND_TYPE', config.get("backend_type", "UNKNOWN")
     )
     backend_type = str(backend_type).upper()
     device_category = os.getenv(
         'DEVICE_CATEGORY', config.get("device_category", "CPU")
     )
-    if 'CUDA_VISIBLE_DEVICES' in os.environ:
+
+    # NPU>GPU>CPU
+    if device_category == "ASCEND":
```
Comment (JoeyHwong-gk) [inline lib/sedna/backend/__init__.py:40]: Maybe it's better to change "ASCEND" to "NPU".

File: lib/sedna/backend/mindspore/__init__.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+            self.estimator.load_weights(model_path)
+
+    def get_weights(self):
+        """todo: no support yet"""
+
+    def set_weights(self, weights):
+        """todo: no support yet"""
+
+    def model_info(self, model, relpath=None, result=None):
+        _, _type = os.path.splitext(model)
```
Comment (JoeyHwong-gk) [inline lib/sedna/backend/mindspore/__init__.py:79]: It seems to be the same as the parent class, suggest to delete.

File: lib/sedna/backend/__init__.py (hunk 1 above)
Comment (chou-shun) [reply lib/sedna/backend/__init__.py:40]: This is because mindspore takes “ASCEND" as "device_target", is it necessary to modify it?
[https://www.mindspore.cn/docs/api/zh-CN/r1.3/api_python/mindspore.context.html?highlight=Ascend#module-mindspore.context](url)

File: lib/sedna/backend/mindspore/__init__.py (hunk 2 above)
Comment (chou-shun) [reply lib/sedna/backend/mindspore/__init__.py:79]: Got it, thanks.

File: lib/sedna/backend/__init__.py (hunk 1 above)
Comment (JoeyHwong-gk) [reply lib/sedna/backend/__init__.py:40]: In fact, this variable `device_category` does not affect the framework, only use to guide our contributors to distinguish between processing units, as usually juxtapose NPU with CPU/GPU, "ASCEND" is only provided by huawei.org. 

In the current scenario, if `MindSpore` supports `ASCEND` only, I think it is best not to set the limit here.

**Comment by chou-shun** [comment]
@JoeyHwong-gk I have revised it according to your opinion. Please review it again. Thanks:)

**Comment by JoeyHwong-gk** [comment]
> @JoeyHwong-gk I have revised it according to your opinion. Please review it again. Thanks:)

thx for your job, it looks good to me

File: lib/sedna/backend/mindspore/__init__.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
+# Copyright 2021 The KubeEdge Authors.
```
Comment (ugvddm) [inline lib/sedna/backend/mindspore/__init__.py:1]: thanks for your great job! BTW, should we provide some example about mindspore to end user? joint inference or incremental learning?

Comment (chou-shun) [reply lib/sedna/backend/mindspore/__init__.py:1]: OK, I'll upload a resnet example by tomorrow

Comment (JimmyYang20) [reply lib/sedna/backend/mindspore/__init__.py:1]: @chou-shun please put your example in `lib/examples/backend/mindspore/`.

File: lib/sedna/backend/mindspore/__init__.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+            context.set_context(mode=context.GRAPH_MODE, device_target="Ascend")
+        elif self.use_cuda:
+            context.set_context(mode=context.GRAPH_MODE, device_target="GPU")
+        else:
+            context.set_context(mode=context.GRAPH_MODE, device_target="CPU")
+
+        if callable(self.estimator):
+            self.estimator = self.estimator()
+
+    def train(self, train_data, valid_data=None, **kwargs):
```
[the remark this answers is not in the document]
Comment (chou-shun) [reply lib/sedna/backend/mindspore/__init__.py:39]: I tried your method, but it didn't work very well. That is because, for class `mindspore.Model`, the data passed to it MUST BE `mindspore.Dataset'[https://www.mindspore.cn/docs/api/zh-CN/r1.3/api_python/mindspore.html?highlight=train_dataset#mindspore.Model.train](url), and can not be a BaseDataSource.
Here are some solutions I thought of：
1. Abandon class `mindspore.Model` and adopt a more flexible approach: `for data in dataloader: loss = network(data), loss.update()`.
2. Repackage `train_data` into a `mindspore.Dataset` in `interface.py`.
For solution 1, it increases the difficulty of model development and is not a popular way.
For solution 2, actually it is counterintuitive.
Which do you prefer? Or do you have a more reasonable solution?

[the remark this answers is not in the document]
Comment (chou-shun) [reply lib/sedna/backend/mindspore/__init__.py:39]: I tried your method, but it didn't work very well. That is because, for method `mindspore.Model.train`, the data passed to it MUST BE a `mindspore.Dataset', and can not be a `BaseDataSource`.
Here are some solutions I thought of：
1. Abandon class `mindspore.Model` and adopt a more flexible approach: `for data in dataloader: loss = network(data), loss.update()`.
2. Repackage `train_data` into a `mindspore.Dataset` in `interface.py`.
For solution 1, it increases the difficulty of model development and is not a popular way.
For solution 2, actually it is counterintuitive.
Which do you prefer? Or do you have a more reasonable solution?

[the remark this answers is not in the document]
Comment (chou-shun) [reply lib/sedna/backend/mindspore/__init__.py:39]: I tried your method, but it didn't work very well. That is because, for class `mindspore.Model`, the data passed to it MUST BE `mindspore.Dataset'[https://www.mindspore.cn/docs/api/zh-CN/r1.3/api_python/mindspore.html?highlight=train_dataset#mindspore.Model.train](url), and can not be a BaseDataSource.
Here are some solutions I thought of：

- 1. Abandon class `mindspore.Model` and adopt a more flexible approach: `for data in dataloader: loss = network(data), loss.update()`.
- 2. Repackage `train_data` into a `mindspore.Dataset` in `interface.py`.

For solution 1, it increases the difficulty of model development and is not a popular way.
For solution 2, actually it is counterintuitive.
Which do you prefer? Or do you have a more reasonable solution?

[the remark this answers is not in the document]
Comment (chou-shun) [reply lib/sedna/backend/mindspore/__init__.py:39]: I tried your method, but it didn't work very well. That is because, for method `mindspore.Model.train`, the data passed to it MUST BE a `mindspore.Dataset', and can not be a `BaseDataSource`.
Here are some solutions I thought of：

- 1. Abandon class `mindspore.Model` and adopt a more flexible approach: `for data in dataloader: loss = network(data), loss.update()`.
- 2. Repackage `train_data` into a `mindspore.Dataset` in `interface.py`.

For solution 1, it increases the difficulty of model development and is not a popular way.
For solution 2, actually it is counterintuitive.
Which do you prefer? Or do you have a more reasonable solution?

[the remark this answers is not in the document]
Comment (chou-shun) [reply lib/sedna/backend/mindspore/__init__.py:39]: I tried your method, but it didn't work very well. That is because, for method `mindspore.Model.train`, the data passed to it MUST BE a `mindspore.Dataset`, and can not be a `BaseDataSource`.
Here are some solutions I thought of：

- 1. Abandon class `mindspore.Model` and adopt a more flexible approach: `for data in dataloader: loss = network(data), loss.update()`.
- 2. Repackage `train_data` into a `mindspore.Dataset` in `interface.py`.

For solution 1, it increases the difficulty of model development and is not a popular way.
For solution 2, actually it is counterintuitive.
Which do you prefer? Or do you have a more reasonable solution?

[the remark this answers is not in the document]
Comment (JoeyHwong-gk) [reply lib/sedna/backend/mindspore/__init__.py:39]: PTAL @jaypume

File: examples/lib-samples/backend/mindspore/ResNet50/README.md [hunk 5]
```diff
@@ -1,70 +1,100 @@
...
+  ├── README.md
+  ├── scripts
+    ├── run_eval.sh                        # launch ascend evaluation
+    ├── run_eval_cpu.sh                    # launch cpu evaluation
+    ├── run_infer.sh                       # launch cpu inference
+    ├── run_standalone_train.sh            # launch ascend standalone training
+    ├── run_standalone_train_cpu.sh        # launch cpu training
+  ├── src
+    ├── config.py                          # parameter configuration
+    ├── dataset.py                         # data preprocessing
```
Comment (ugvddm) [inline examples/lib-samples/backend/mindspore/ResNet50/README.md:18]: should we move the way of data preprocessing for midnspore to sedna lib,  instead of in examples?  the other people may be want to use it when they develop a appliation based on mindspore.

Comment (JoeyHwong-gk) [reply examples/lib-samples/backend/mindspore/ResNet50/README.md:18]: Where do you think is the most appropriate ? `sedna.datasource` or `sedna.backend.mindspore` ?

Comment (chou-shun) [reply examples/lib-samples/backend/mindspore/ResNet50/README.md:18]: Should we develop preprocessing methods **for each model**, or develop a **general** preprocessing method?
In my assumption, for a certain scene (such as image classification), we can predefine several fixed preprocessing methods (such as normalize, resize), and the user only needs to pass some parameters.

File: lib/sedna/backend/mindspore/__init__.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jaypume) [reply lib/sedna/backend/mindspore/__init__.py:39]: I suggest that `BaseDataSorce` is designed to compatible with existing `Dataset` such as `mindspore.Dataset`, `torch.utils.data.Dataset`, `tf.data.Dataset`. If the Dataset instance is passed to for example `FederatedLearning.train(tfDatasetInstance)`, it should work.

File: examples/lib-samples/backend/mindspore/ResNet50/README.md (hunk 5 above)
Comment (jaypume) [reply examples/lib-samples/backend/mindspore/ResNet50/README.md:18]: I suggest not to integrate data preprocessing functions into Sedna Lib, for the reason:
- There are various preprocessing functions, and they cannot be covered by Sedna.
- The data preprocessing functions are always already defined in developer's training code or AI framework.

File: lib/sedna/backend/mindspore/__init__.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (JoeyHwong-gk) [reply lib/sedna/backend/mindspore/__init__.py:39]: > I suggest that `BaseDataSorce` is designed to compatible with existing `Dataset` such as `mindspore.Dataset`, `torch.utils.data.Dataset`, `tf.data.Dataset`. If the Dataset instance is passed to for example `FederatedLearning.train(tfDatasetInstance)`, it should work.

You can help us raise an issue so we can track it.
